File: languages/typescript/projects/02-core/main.ts
import * as fs from "node:fs";

type ScoreRecord = {
    name: string;
    score: number;
};

function parseRow(line: string): ScoreRecord | null {
    const parts = line.trim().split(/\s+/);
    if (parts.length < 2) {
        return null;
    }

    const score = Number.parseInt(parts.at(-1) ?? "", 10);
    if (!Number.isInteger(score) || score < 0 || score > 100) {
        return null;
    }

    const name = parts.slice(0, -1).join(" ");
    if (!name) {
        return null;
    }

    return { name, score };
}

function main(): void {
    const inputPath = fs.readFileSync(0, "utf8").trim();
    if (!inputPath) {
        console.log("Expected an input file path.");
        return;
    }
    if (!fs.existsSync(inputPath)) {
        console.log("Input file not found.");
        return;
    }

    const validRecords: ScoreRecord[] = [];
    let invalidRows = 0;

    for (const line of fs.readFileSync(inputPath, "utf8").split(/\r?\n/)) {
        if (line.trim().length === 0) {
            continue;
        }

        const record = parseRow(line);
        if (record === null) {
            invalidRows += 1;
            continue;
        }

        validRecords.push(record);
    }

    const reportLines = [
        "Grade Report",
        `Source file: ${inputPath}`,
        `Valid records: ${validRecords.length}`,
        `Invalid rows skipped: ${invalidRows}`,
    ];

    if (validRecords.length === 0) {
        reportLines.push(
            "Average: N/A",
            "Minimum: N/A",
            "Maximum: N/A",
            "Students: none",
        );
        console.log("Valid records: 0");
        console.log(`Invalid rows skipped: ${invalidRows}`);
        console.log("Average: N/A");
    } else {
        const total = validRecords.reduce(
            (sum, record) => sum + record.score,
            0,
        );
        const minimum = validRecords.reduce(
            (best, record) => Math.min(best, record.score),
            validRecords[0]!.score,
        );
        const maximum = validRecords.reduce(
            (best, record) => Math.max(best, record.score),
            validRecords[0]!.score,
        );
        reportLines.push(
            `Average: ${(total / validRecords.length).toFixed(2)}`,
            `Minimum: ${minimum}`,
            `Maximum: ${maximum}`,
            "Students:",
        );
        for (const record of validRecords) {
            reportLines.push(`- ${record.name}: ${record.score}`);
        }

        console.log(`Valid records: ${validRecords.length}`);
        console.log(`Invalid rows skipped: ${invalidRows}`);
        console.log(`Average: ${(total / validRecords.length).toFixed(2)}`);
    }

    fs.writeFileSync("report.txt", reportLines.join("\n") + "\n", "utf8");
    console.log("Report written to report.txt");
}

main();

File: languages/java/projects/02-core/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Scanner;

public class Main {
    record StudentScore(String name, int score) {
    }

    static StudentScore parseRecord(String line) {
        String[] parts = line.trim().split("\\s+");
        if (parts.length < 3) {
            return null;
        }

        try {
            int score = Integer.parseInt(parts[parts.length - 1]);
            if (score < 0 || score > 100) {
                return null;
            }
            String name = String.join(" ", java.util.Arrays.copyOf(parts, parts.length - 1));
            return name.isBlank() ? null : new StudentScore(name, score);
        } catch (NumberFormatException error) {
            return null;
        }
    }

    public static void main(String[] args) throws IOException {
        Locale.setDefault(Locale.US);
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the input file path: ");
        Path sourcePath = Path.of(scanner.nextLine().trim());

        if (!Files.isRegularFile(sourcePath)) {
            System.out.println("Input file not found.");
            return;
        }

        List<StudentScore> validRecords = new ArrayList<>();
        int invalidRows = 0;

        for (String line : Files.readAllLines(sourcePath)) {
            if (line.isBlank()) {
                continue;
            }

            StudentScore record = parseRecord(line);
            if (record == null) {
                invalidRows++;
            } else {
                validRecords.add(record);
            }
        }

        Path reportPath = Path.of("report.txt");
        List<String> lines = new ArrayList<>();
        lines.add("Grade Report");
        lines.add("Source file: " + sourcePath);
        lines.add("Valid records: " + validRecords.size());
        lines.add("Invalid rows skipped: " + invalidRows);

        if (validRecords.isEmpty()) {
            lines.add("Average: N/A");
            lines.add("Students: none");
            System.out.println("Valid records: 0");
            System.out.println("Invalid rows skipped: " + invalidRows);
            System.out.println("Average: N/A");
        } else {
            double average = validRecords.stream().mapToInt(StudentScore::score).average().orElse(0.0);
            lines.add(String.format("Average: %.2f", average));
            lines.add("Students:");
            for (StudentScore record : validRecords) {
                lines.add("- " + record.name() + ": " + record.score());
            }
            System.out.println("Valid records: " + validRecords.size());
            System.out.println("Invalid rows skipped: " + invalidRows);
            System.out.printf("Average: %.2f%n", average);
        }

        Files.write(reportPath, lines);
        System.out.println("Report written to report.txt");
    }
}
